redbox-core/redbox/models/chain.py

```diff
@@ -0,0 +1,280 @@
+
+
+class RedboxQuery(BaseModel):
+    question: str = Field(description="The last user chat message")
+    s3_keys: list[str] = Field(description="List of files to process", default_factory=list)
+    user_uuid: UUID = Field(description="User the chain in executing for")
+    chat_history: list[ChainChatMessage] = Field(description="All previous messages in chat (excluding question)")
+    ai_settings: AISettings = Field(description="User request AI settings", default_factory=AISettings)
+    permitted_s3_keys: list[str] = Field(description="List of permitted files for response", default_factory=list)
+
+
+class LLMCallMetadata(BaseModel):
+    id: str = Field(default_factory=lambda: str(uuid4()))
+    llm_model_name: str
+    input_tokens: int
+    output_tokens: int
+    timestamp: datetime = Field(default_factory=lambda: datetime.now(UTC))
+
+    model_config = {"frozen": True}
+
+
+class RequestMetadata(BaseModel):
+    llm_calls: list[LLMCallMetadata] = Field(default_factory=list)
+    selected_files_total_tokens: int = 0
+    number_of_selected_files: int = 0
+
+    @property
+    def input_tokens(self) -> dict[str, int]:
+        """
+        Creates a dictionary of model names to number of input tokens used
+        """
+        tokens_by_model = dict()
+        for call_metadata in self.llm_calls:
+            tokens_by_model[call_metadata.llm_model_name] = (
+                tokens_by_model.get(call_metadata.llm_model_name, 0) + call_metadata.input_tokens
+            )
+        return tokens_by_model
+
+    @property
+    def output_tokens(self):
+        """
+        Creates a dictionary of model names to number of output tokens used
+        """
+        tokens_by_model = dict()
+        for call_metadata in self.llm_calls:
+            tokens_by_model[call_metadata.llm_model_name] = (
+                tokens_by_model.get(call_metadata.llm_model_name, 0) + call_metadata.output_tokens
+            )
+        return tokens_by_model
+
+
+def metadata_reducer(
+    current: RequestMetadata | None,
+    update: RequestMetadata | list[RequestMetadata] | None,
+):
+    """Merges two metadata states."""
+    # If update is actually a list of state updates, run them one by one
+    if isinstance(update, list):
+        reduced = reduce(lambda current, update: metadata_reducer(current, update), update, current)
+        return reduced
+
+    if current is None:
+        return update
+    if update is None:
+        return current
+
+    return RequestMetadata(
+        llm_calls=sorted(set(current.llm_calls) | set(update.llm_calls), key=lambda c: c.timestamp),
+        selected_files_total_tokens=update.selected_files_total_tokens or current.selected_files_total_tokens,
+        number_of_selected_files=update.number_of_selected_files or current.number_of_selected_files,
+    )
+
+
+class RedboxState(BaseModel):
+    request: RedboxQuery
+    documents: Annotated[DocumentState, document_reducer] = DocumentState()
```

Comment:
Are we still going to attatch documents to the state like this? Or have it be something that is more linked to the chat history?

Reply:
TBC I think? 
* do we want the docs to be ordered? 
* so we want the docs to be part of the conversation so that it makes sense for a user to talk about "my last document"?
* how do we handle the case when the total doc size exceede that of the context window? 

my gut feeling is that we do want the docs to be part of the chat history but good to make this clear (if indeed it is the case)?

Reply:
I have the same gut feeling. I don't know how well the model deals with a situation that looks like: 

User: Question about a document [attaches doc]
AI: Answer about a document 
User: Question about different document [attaches doc]
AI: Answer about a document 

If the documents are only attached in the latest prompt and not in the rest of the history?

Reply:
But this feels like a big change that needs to be thought through. Would love @jamesrichards4 view on it